pkg/querier/worker/scheduler_processor.go

```diff
@@ -111,42 +119,67 @@ func (sp *schedulerProcessor) processQueriesOnSingleStream(workerCtx context.Con
 	for backoff.Ongoing() {
-		c, err := schedulerClient.QuerierLoop(execCtx)
-		if err == nil {
-			err = c.Send(&schedulerpb.QuerierToScheduler{QuerierID: sp.querierID})
-		}
-
-		if err != nil {
-			level.Warn(sp.log).Log("msg", "error contacting scheduler", "err", err, "addr", address)
-			backoff.Wait()
-			continue
-		}
-
-		if err := sp.querierLoop(c, address, inflightQuery); err != nil {
-			// Do not log an error is the query-scheduler is shutting down.
-			if s, ok := status.FromError(err); !ok ||
-				(!strings.Contains(s.Message(), schedulerpb.ErrSchedulerIsNotRunning.Error()) &&
-					!strings.Contains(s.Message(), context.DeadlineExceeded.Error()) &&
-					!strings.Contains(s.Message(), "stream terminated")) {
-				level.Error(sp.log).Log("msg", "error processing requests from scheduler", "err", err, "addr", address)
-			}
-			if strings.Contains(err.Error(), context.DeadlineExceeded.Error()) || strings.Contains(err.Error(), "stream terminated") {
-				backoff.Reset()
-				continue
+		func() {
+			if err := sp.querierLoop(execCtx, schedulerClient, address, inflightQuery); err != nil {
+				// Do not log an error is the query-scheduler is shutting down.
+				if s, ok := status.FromError(err); !ok ||
+					(!strings.Contains(s.Message(), schedulerpb.ErrSchedulerIsNotRunning.Error()) &&
+						!strings.Contains(s.Message(), context.Canceled.Error()) &&
+						!strings.Contains(s.Message(), "stream terminated")) {
+					level.Error(sp.log).Log("msg", "error processing requests from scheduler", "err", err, "addr", address)
+				}
+				if strings.Contains(err.Error(), context.Canceled.Error()) || strings.Contains(err.Error(), "stream terminated") {
+					backoff.Reset()
+					return
+				}
+				backoff.Wait()
+				return
 			}
-			backoff.Wait()
-			continue
-		}
 
-		backoff.Reset()
+			backoff.Reset()
+		}()
 	}
 }
 
 // process loops processing requests on an established stream.
-func (sp *schedulerProcessor) querierLoop(c schedulerpb.SchedulerForQuerier_QuerierLoopClient, address string, inflightQuery *atomic.Bool) error {
-	// Build a child context so we can cancel a query when the stream is closed.
-	ctx, cancel := context.WithCancel(c.Context())
-	defer cancel()
+func (sp *schedulerProcessor) querierLoop(parentCtx context.Context, schedulerClient schedulerpb.SchedulerForQuerierClient, address string, inflightQuery *atomic.Bool) error {
+	loopCtx, loopCancel := context.WithCancel(parentCtx)
+	defer loopCancel()
+
+	if sp.maxLoopDuration > 0 {
+		go func() {
+			timer := time.NewTimer(jitter(sp.maxLoopDuration, 0.3))
+			defer timer.Stop()
+
+			select {
+			case <-timer.C:
+				level.Debug(sp.log).Log("msg", "waiting for inflight queries to complete")
+				for inflightQuery.Load() {
+					select {
+					case <-parentCtx.Done():
+						// In the meanwhile, the execution context has been explicitly canceled, so we should just terminate.
+						return
+					default:
+						time.Sleep(100 * time.Millisecond)
```

Comment:
You mind commenting why this is neccessary?

```suggestion
						time.Sleep(100 * time.Millisecond)
```